### PR TITLE
Visco elastic notebook

### PR DESCRIPTION
**Pull request by rhodrin**
Add a notebook for a viscoelastic model. Add elastic and viscoelastic options to benchmark.py.

### CODE REVIEW

File: examples/seismic/viscoelastic/viscoelastic_example.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        space_order=4, nbpml=40, autotune=False, constant=False, **kwargs):
+
+    solver = viscoelastic_setup(shape=shape, spacing=spacing, nbpml=nbpml, tn=tn,
+                                space_order=space_order, constant=constant, **kwargs)
+    info("Applying Forward")
+    # Define receiver geometry (spread across x, just below surface)
+    rec1, rec2, vx, vz, txx, tzz, txz, rxx, rzz, rxz, summary = \
+        solver.forward(autotune=autotune)
+
+    return rec1, rec2, vx, vz, txx, tzz, txz, rxx, rzz, rxz, summary
```
Comment (mloubout) [inline examples/seismic/viscoelastic/viscoelastic_example.py]: Nitpicking, but the `r` variables are auxiliary fields to model the attenuation Q, noone really cares what they look like so don't need to ouptut it or anything. Can also ignore it in `wavesolver` and let the operator use the ones created in `operator`, should make it easier to follow, but fine with me like that if you want it.

File: examples/seismic/viscoelastic/operators.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+                           time_order=1, space_order=space_order)
+        txy = TimeFunction(name='txy', grid=model.grid, staggered=stagg_xy, save=save,
+                           time_order=1, space_order=space_order)
+        tyz = TimeFunction(name='tyz', grid=model.grid, staggered=stagg_yz, save=save,
+                           time_order=1, space_order=space_order)
+
+    return txx, tyy, tzz, txy, txz, tyz
+
+
+def memory_variables(model, save, space_order):
```
Comment (mloubout) [inline examples/seismic/viscoelastic/operators.py]: auxiliary/attenuation variables, memory is a weird name for it.

File: examples/seismic/viscoelastic/operators.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                           time_order=1, space_order=space_order)
+        rxy = TimeFunction(name='rxy', grid=model.grid, staggered=stagg_xy, save=save,
+                           time_order=1, space_order=space_order)
+        ryz = TimeFunction(name='ryz', grid=model.grid, staggered=stagg_yz, save=save,
+                           time_order=1, space_order=space_order)
+
+    return rxx, ryy, rzz, rxy, rxz, ryz
+
+
+def particle_velocity_fields(model, save, space_order):
```
Comment (mloubout) [inline examples/seismic/viscoelastic/operators.py]: import from elastic, this is copy paste

File: examples/seismic/viscoelastic/operators.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
+import numpy as np
+import sympy as sp
+
+from devito import Eq, Operator, TimeFunction, NODE
+from examples.seismic import PointSource, Receiver
+
+
+def stress_fields(model, save, space_order):
```
Comment (mloubout) [inline examples/seismic/viscoelastic/operators.py]: import from elastic

File: examples/seismic/viscoelastic/operators.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+    u_ryz = Eq(ryz.forward, ryz
+               - damp*s*1/t_s*(ryz+mu*(t_es/t_s-1)*(vy.forward.dz+vz.forward.dy)))
+
+    src_rec_expr = src_rec(vx, vy, vz, txx, tyy, tzz, rxx, ryy, rzz, model, geometry)
+    return [u_vx, u_vy, u_vz, u_txx, u_tyy, u_tzz, u_txz, u_txy, u_tyz,
+            u_rxx, u_ryy, u_rzz, u_rxz, u_rxy, u_ryz] + src_rec_expr
+
+
+def src_rec(vx, vy, vz, txx, tyy, tzz, rxx, ryy, rzz, model, geometry):
```
Comment (mloubout) [inline examples/seismic/viscoelastic/operators.py]: import from elastic

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 6]
```diff
@@ -0,0 +1,338 @@
+{
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: Overall neeeds bit more explanation/comments/... a bit too brute force like that.
Maybe split the equations in different cells for velocities/stress/attenuation

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    "\n",
+    "# Model physical parameters:\n",
+    "vp = np.zeros(shape)\n",
+    "qp = np.zeros(shape)\n",
+    "vs = np.zeros(shape)\n",
+    "qs = np.zeros(shape)\n",
+    "rho = np.zeros(shape)\n",
+    "\n",
+    "# Set up three horizontally separated layers:\n",
+    "for i in range(shape[0]):\n",
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: This is very intricated for not much, ie for vp 
`vp = 1.52 * np.ones(shape);vp[:,:,0.5*extent:] = 1.6;vp[:,:,0.5*extent+4h:] = 2.2; `

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Domain size:\n",
+    "extent = (200., 100., 100.)\n",
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: `(201, 101, 101)` will give you nicer spacing value as currently it's 200/199 which is not great.

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    "\n",
+    "This is a first attempt at implemenenting the viscoelastic wave equation as described in:\n",
+    "\n",
+    "[1] Johan O. A. Roberston, *et.al.* (1994). \"Viscoelatic finite-difference modeling\" GEOPHYSICS, 59(9), 1444-1456.\n",
+    "\n",
+    "See also the FDELMODC implementation by Jan Thorbecke:\n",
+    "\n",
+    "[2] https://janth.home.xs4all.nl/Software/fdelmodcManual.pdf\n",
+    "\n",
+    "In the following example, a three dimensional toy problem will be introduced consiting of a single Ricker source located at (100, 50, 35) in a 200 x 100 x 100 km domain."
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: 200km :), oil companies will love this :D

File: examples/seismic/viscoelastic/operators.py (hunk 2 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/operators.py]: Some people call them memory variables i.e. Bohlen (2002) among others. Anything, within reason, is fine with me though.

File: examples/seismic/viscoelastic/viscoelastic_example.py (hunk 1 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/viscoelastic_example.py]: Sure, could tidy that up.

File: examples/seismic/tutorials/09_viscoelastic.ipynb (hunk 7 above)
Comment (rhodrin) [reply examples/seismic/tutorials/09_viscoelastic.ipynb]: Yeah.

File: examples/seismic/tutorials/09_viscoelastic.ipynb (hunk 8 above)
Comment (rhodrin) [reply examples/seismic/tutorials/09_viscoelastic.ipynb]: The spacing is currently (1, 1, 1)?

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+    "# Source freq. in MHz (note that the source is defined below):\n",
+    "f0 = 0.12\n",
+    "\n",
+    "# Thorbecke's parameter notation\n",
+    "cp2 = model.vp*model.vp\n",
+    "cs2 = model.vs*model.vs\n",
+    "ro = 1./model.rho\n",
+    "\n",
+    "mu = cs2*ro\n",
+    "l = (cp2*ro - 2*mu)\n",
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: These are wrong, saw that doing vector branch, correct parameters are
```
ro = 1/rho
mu = cs2 * rho
l = (cp2*rho - 2*mu
```
it is the density for lambda/rho not its invers

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+    "        - 2*mu*(t_es/t_s-1)*(vx.forward.dx+vy.forward.dy)))\n",
+    "eq_rzz = Eq(rzz.forward, damp*rzz \\\n",
+    "    - damp*s*1./t_s*(rzz+pi*(t_ep/t_s-1)*(vx.forward.dx+vy.forward.dy+vz.forward.dz) \\\n",
+    "        - 2*mu*(t_es/t_s-1)*(vx.forward.dx+vy.forward.dy)))\n",
+    "eq_rxy = Eq(rxy.forward, damp*rxy \\\n",
+    "    - damp*s*1/t_s*(rxy+mu*(t_es/t_s-1)*(vx.forward.dy+vy.forward.dx)))\n",
+    "eq_rxz = Eq(rxz.forward, damp*rxz \\\n",
+    "    - damp*s*1/t_s*(rxz+mu*(t_es/t_s-1)*(vx.forward.dz+vz.forward.dx)))\n",
+    "eq_ryz = Eq(ryz.forward, ryz \\\n",
+    "    - damp*s*1/t_s*(ryz+mu*(t_es/t_s-1)*(vy.forward.dz+vz.forward.dy)))"
```
Comment (mloubout) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: Double check equations too there was a sign error in one of the elastic notebook.

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+    "vx.data.shape"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since our functions are first order in time, the time dimension is of length 2. The spatial extent of the data includes the absorbing boundary layers in each dimension.\n",
+    "\n",
+    "Aslo, let us look at:"
```
Comment (georgebisbas) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: Also*

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 13]
```diff
@@ -0,0 +1,465 @@
...
+   "source": [
+    "# Execute the operator. NOTE: This may take up to several minutes.\n",
+    "op(dt=dt)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before plotting some results, let us first look at the shape of the data stoted in one of our time functions:"
```
Comment (georgebisbas) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: stored*

File: examples/seismic/tutorials/09_viscoelastic.ipynb [hunk 14]
```diff
@@ -0,0 +1,465 @@
...
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Operator `Kernel` run in 67.81 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Execute the operator. NOTE: This may take up to several minutes.\n",
```
Comment (georgebisbas) [inline examples/seismic/tutorials/09_viscoelastic.ipynb]: Maybe not necessary? With no optimizations no openmp/mpi I got 115 secs on a very weak laptop.

File: benchmarks/user/benchmark.py [hunk 15]
```diff
@@ -110,7 +114,14 @@ def run(problem, **kwargs):
     """
     A single run with a specific set of performance parameters.
     """
-    setup = tti_setup if problem == 'tti' else acoustic_setup
+    if problem == 'viscoelastic':
```
Comment (FabioLuporini) [inline benchmarks/user/benchmark.py]: maybe we start needing a global dict

```
{'viscoelastic': viscoelastic_setup,
 'elastic': ...,
 ...}
```

File: examples/seismic/model.py [hunk 16]
```diff
@@ -722,4 +791,98 @@ def critical_dt(self):
...
+    def __init__(self, origin, spacing, shape, space_order, vp, qp, vs, qs, rho,
+                 nbpml=20, dtype=np.float32):
+        super(ModelViscoelastic, self).__init__(origin, spacing, shape, space_order,
+                                                nbpml=nbpml, dtype=dtype)
+        # Create dampening field as symbol `damp`
+        self.damp = Function(name="damp", grid=self.grid)
+        initialize_damp(self.damp, self.nbpml, self.spacing, mask=True)
+
+        # Create square slowness of the wave as symbol `m`
+        if isinstance(vp, np.ndarray):
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: these 6-7 lines... shouldn't they become a function?

File: examples/seismic/model.py [hunk 17]
```diff
@@ -722,4 +791,98 @@ def critical_dt(self):
...
+            self.vp = Function(name="vp", grid=self.grid, space_order=space_order)
+            initialize_function(self.vp, vp, self.nbpml)
+        else:
+            self.vp = Constant(name="vp", value=vp)
+        self._physical_parameters = ('vp',)
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: nitpicking: `physical_parameters` should be a list and then finally turned into a tuple and assigned to ``self. `

File: examples/seismic/model.py [hunk 18]
```diff
@@ -722,4 +797,99 @@ def critical_dt(self):
         #
         # The CFL condtion is then given by
         # dt < h / (sqrt(2) * max(vp)))
-        return self.dtype(.5*mmin(self.spacing) / (np.sqrt(2)*mmax(self.vp)))
+        # FIXME: Fix 'Constant' so that that mmax(self.vp) returns the data value
+        return self.dtype(.5*mmin(self.spacing) / (np.sqrt(2)*mmax(self.vp.data)))
+
+
+class ModelViscoelastic(GenericModel):
```
Comment (mloubout) [inline examples/seismic/model.py]: Wouldn't it be easier to inherit from Elastic model and only setup the extras parameters?

File: examples/seismic/model.py (hunk 16 above)
Comment (mloubout) [reply examples/seismic/model.py]: Probably would be easier I guess

File: examples/seismic/tutorials/06_elastic.ipynb [hunk 19]
```diff
@@ -23,7 +23,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
+   "execution_count": 11,
```
Comment (mloubout) [inline examples/seismic/tutorials/06_elastic.ipynb]: Haven't looked at ti, but why is this notebook changing?

File: examples/seismic/viscoelastic/operators.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
+import numpy as np
+import sympy as sp
+
+from devito import Eq, Operator, TimeFunction, NODE
+from examples.seismic.elastic import stress_fields, particle_velocity_fields, src_rec
+
+
+def memory_variables(model, save, space_order):
```
Comment (mloubout) [inline examples/seismic/viscoelastic/operators.py]: This looks like it's the exact same setup as the stress tensor. Maybe a single function as tensor_function(name,...)?

Will be sing TensorFunction after my PR anyway but would be less work to update :D

File: examples/seismic/viscoelastic/viscoelastic_example.py [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+        src_coordinates[0, -1] = model.origin[-1] + 2 * spacing[-1]
+
+    rec_coordinates = np.empty((nrec, len(spacing)))
+    rec_coordinates[:, 0] = np.linspace(0., model.domain_size[0], num=nrec)
+    if len(shape) > 1:
+        rec_coordinates[:, 1] = np.array(model.domain_size)[1] * .5
+        rec_coordinates[:, -1] = model.origin[-1] + 2 * spacing[-1]
+    geometry = AcquisitionGeometry(model, rec_coordinates, src_coordinates,
+                                   t0=0.0, tn=tn, src_type='Ricker', f0=0.12)
+
```
Comment (mloubout) [inline examples/seismic/viscoelastic/viscoelastic_example.py]: We may want to make a `setup_gemoetry` in `utils.py` or something as this is the same setup code in all the seismic examples.

File: examples/seismic/model.py (hunk 18 above)
Comment (rhodrin) [reply examples/seismic/model.py]: Yeah, was thinking about doing it that way earlier. Will try and remember if there's a reason I didn't immediately do that.

File: examples/seismic/tutorials/06_elastic.ipynb (hunk 19 above)
Comment (rhodrin) [reply examples/seismic/tutorials/06_elastic.ipynb]: Yeah, that's been annoying me. Will try and figure out why exactly.

File: examples/seismic/viscoelastic/operators.py (hunk 20 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/operators.py]: Sure, can do.

File: examples/seismic/model.py [hunk 22]
```diff
@@ -55,6 +56,25 @@ def demo_model(preset, **kwargs):
                             shape=shape, dtype=dtype, spacing=spacing, nbpml=nbpml,
                             **kwargs)
 
+    if preset.lower() in ['constant-viscoelastic']:
+        # A constant single-layer model in a 2D or 3D domain
+        # with velocity 2.2km/s.
```
Comment (georgebisbas) [inline examples/seismic/model.py]: Should the comments reflect a constant value?
What I mean is that maybe it is better to have:

        # A constant single-layer model in a 2D or 3D domain
        # with velocity `vp` km/s.

File: examples/seismic/model.py [hunk 23]
```diff
@@ -722,4 +793,68 @@ def critical_dt(self):
...
+        physical_parameters.append('qs')
+
+        self._physical_parameters = as_tuple(physical_parameters)
+
+    @property
+    def critical_dt(self):
+        """
+        Critical computational time step value from the CFL condition.
+        """
+        # For a fixed time order this number goes down as the space order increases.
```
Comment (georgebisbas) [inline examples/seismic/model.py]: this number decreases*?

File: examples/seismic/viscoelastic/wavesolver.py [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+    Notes
+    -----
+    This is an experimental staggered grid viscoelastic modeling kernel.
+    """
+    def __init__(self, model, geometry, space_order=4, **kwargs):
+        self.model = model
+        self.geometry = geometry
+
+        self.space_order = space_order
+        # Time step can be \sqrt{3}=1.73 bigger with 4th order
```
Comment (georgebisbas) [inline examples/seismic/viscoelastic/wavesolver.py]: Maybe revisit this relation, `bigger than what`?

File: examples/seismic/viscoelastic/wavesolver.py [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+        kwargs['rxz'] = rxz
+        if self.model.grid.dim == 3:
+            kwargs['vy'] = vy
+            kwargs['tyy'] = tyy
+            kwargs['txy'] = txy
+            kwargs['tyz'] = tyz
+            kwargs['ryy'] = ryy
+            kwargs['rxy'] = rxy
+            kwargs['ryz'] = ryz
+        # Pick m from model unless explicitly provided
```
Comment (georgebisbas) [inline examples/seismic/viscoelastic/wavesolver.py]: vp* insterad of m?

File: benchmarks/user/benchmark.py (hunk 15 above)
Comment (FabioLuporini) [reply benchmarks/user/benchmark.py]: reminder

File: examples/seismic/model.py [hunk 26]
```diff
@@ -689,29 +763,26 @@ def __init__(self, origin, spacing, shape, space_order, vp, vs, rho, nbpml=20,
...
-            initialize_function(self.rho, rho, self.nbpml)
+        self.vs = self._gen_phys_param(vs, 'vs', space_order)
+        physical_parameters.append('vs')
+
+        self.rho = self._gen_phys_param(rho, 'rho', space_order)
+        physical_parameters.append('rho')
+
+        self._physical_parameters = as_tuple(physical_parameters)
+
+    def _gen_phys_param(self, field, name, space_order):
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: @rhodrin @mloubout not necessarily in this PR, but...can we use this function all over the place please?

File: examples/seismic/model.py [hunk 27]
```diff
@@ -689,29 +763,26 @@ def __init__(self, origin, spacing, shape, space_order, vp, vs, rho, nbpml=20,
...
+        if isinstance(field, np.ndarray):
+            devito_field = Function(name=name, grid=self.grid, space_order=space_order)
```
Comment (FabioLuporini) [inline examples/seismic/model.py]: typically we call these just ``function``

Comment (FabioLuporini) [reply examples/seismic/model.py]: or ``field``

File: examples/seismic/viscoelastic/wavesolver.py (hunk 25 above)
Comment (FabioLuporini) [reply examples/seismic/viscoelastic/wavesolver.py]: and not just `vp`. I think it's copy-paste leftover

File: benchmarks/user/benchmark.py (hunk 15 above)
Comment (rhodrin) [reply benchmarks/user/benchmark.py]: Dict added.

File: examples/seismic/model.py (hunk 22 above)
Comment (rhodrin) [reply examples/seismic/model.py]: Velocity is fine here, people will know you're referring to `vp` if you don't specify.

File: examples/seismic/model.py (hunk 27 above)
Comment (rhodrin) [reply examples/seismic/model.py]: Oki, changed it to `function` since `field` is already present.

File: examples/seismic/model.py (hunk 23 above)
Comment (rhodrin) [reply examples/seismic/model.py]: Indeed.

File: examples/seismic/viscoelastic/viscoelastic_example.py (hunk 21 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/viscoelastic_example.py]: Do we want to move this now or wait for a future update?

File: examples/seismic/viscoelastic/wavesolver.py (hunk 24 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/wavesolver.py]: Comment removed for now.

File: examples/seismic/viscoelastic/wavesolver.py (hunk 25 above)
Comment (rhodrin) [reply examples/seismic/viscoelastic/wavesolver.py]: Yeah, was a copy and paste relic here and in elastic. Changed to 'physical parameters'.

File: examples/seismic/viscoelastic/viscoelastic_example.py (hunk 21 above)
Comment (FabioLuporini) [reply examples/seismic/viscoelastic/viscoelastic_example.py]: can you do this in a separate PR?

I'd really like this

**Comment by FabioLuporini** [comment]
Merged.

**Pull request merged**